File: BraintreeIntegrationTests/Braintree-API-Integration-Specs/BTAPIClient_IntegrationTests.m
#import <BraintreeCore/BraintreeCore.h>
#import <BraintreeCore/BTAPIClient_Internal.h>
#import <XCTest/XCTest.h>

@interface BTAPIClient_IntegrationTests : XCTestCase
@end

@implementation BTAPIClient_IntegrationTests {
    BTAPIClient *client;
}

- (void)setUp {
    [super setUp];
    client = [[BTAPIClient alloc] initWithClientKey:@"development_testing_integration_merchant_id"];
}

- (void)testFetchConfiguration_returnsTheConfiguration {
    XCTestExpectation *expectation = [self expectationWithDescription:@"Fetch configuration"];
    [client fetchOrReturnRemoteConfiguration:^(BTConfiguration *configuration, NSError *error) {
        XCTAssertEqualObjects(configuration.json[@"merchantId"].asString, @"integration_merchant_id");
        XCTAssertNil(error);
        [expectation fulfill];
    }];

    [self waitForExpectationsWithTimeout:5 handler:nil];
}

- (void)testPostAnalytics_whenCalled_isSuccessful {
    XCTestExpectation *expectation = [self expectationWithDescription:@"Post analytics event"];

    // Analytics require an authorization fingerprint, needs support for client key
    NSString *event = @"hello world! 🐴";
    [client sendAnalyticsEvent:event completion:^(NSError *error) {
        XCTAssertNil(error);
        [expectation fulfill];
    }];

    [self waitForExpectationsWithTimeout:5 handler:nil];
}

// Testing that analytics "is successful but does not send the event when analytics URL is omitted from the client token"
// is covered by the unit test:
//   testSendAnalyticsEvent_whenRemoteConfigurationHasNoAnalyticsURL_doesNotSendEvent

@end

File: IntegrationTests/Braintree-API-Integration-Specs/BTAPIClient_IntegrationTests.swift
import XCTest
@testable import BraintreeCore

final class BTAPIClient_IntegrationTests: XCTestCase {

    func testFetchConfiguration_withTokenizationKey_returnsTheConfiguration() async throws {
        let apiClient = BTAPIClient(authorization: BTIntegrationTestsConstants.sandboxTokenizationKey)

        let configuration = try await apiClient.fetchOrReturnRemoteConfiguration()
        XCTAssertEqual(configuration.json?["merchantId"].asString(), "dcpspy2brwdjr3qn")
    }

    func testFetchConfiguration_withClientToken_returnsTheConfiguration() async throws {
        let apiClient = BTAPIClient(authorization: BTIntegrationTestsConstants.sandboxClientToken)

        let configuration = try await apiClient.fetchOrReturnRemoteConfiguration()
        // Note: client token uses a different merchant ID than the merchant whose tokenization key
        // we use in the other test
        XCTAssertEqual(configuration.json?["merchantId"].asString(), "348pk9cgf3bgyw2b")
    }

    func testFetchConfiguration_withVersionThreeClientToken_returnsTheConfiguration() async throws {
        let apiClient = BTAPIClient(authorization: BTIntegrationTestsConstants.sandboxClientTokenVersion3)

        let configuration = try await apiClient.fetchOrReturnRemoteConfiguration()
        // Note: client token uses a different merchant ID than the merchant whose tokenization key
        // we use in the other test
        XCTAssertEqual(configuration.json?["merchantId"].asString(), "dcpspy2brwdjr3qn")
    }
}
